scripts/lite-fhir-schema.js

```diff
@@ -0,0 +1,93 @@
+/* eslint-disable @typescript-eslint/no-var-requires */
+const fs = require("fs");
+const fhir = require("../src/sg/gov/moh/fhir/4.0.1/schema.json");
+
+const filename = "src/sg/gov/moh/fhir/4.0.1/lite-schema.json";
+
+const resources = ["Bundle", "Device", "Observation", "Organization", "Patient", "Practitioner", "Specimen"];
+const resourceQueue = [...resources];
+
+const liteSchema = {
+  $schema: "http://json-schema.org/draft-07/schema#",
+  $id: "https://schemata.openattestation.com/sg/gov/moh/fhir/4.0.1/lite-schema.json#",
+  // id: "http://hl7.org/fhir/json-schema/4.0",
+  title: "Lite FHIR schema for Notarise.gov.sg HealthCerts",
+  description: "see http://hl7.org/fhir/json.html#schema for information about the FHIR Json Schemas",
+  $ref: "#/definitions/Bundle",
+  definitions: {}
+};
+
+/** ====== Helper Functions ====== **/
+
+function stringToRef(ref = "") {
+  return ref.replace("#/definitions/", "");
+}
+
+function findUniqueNestedReferences(obj = {}, searchKey = "$ref", results = []) {
+  const keys = Object.keys(obj);
+  const references = results;
+
+  for (const key of keys) {
+    const value = obj[key];
+    if (key === searchKey && typeof value !== "object") {
+      references.push(stringToRef(value));
+    } else if (typeof value === "object") {
+      findUniqueNestedReferences(value, searchKey, references);
+    }
+  }
+
+  return [...new Set(references)];
+}
+
+/** ====== Main Script ====== **/
+
+while (resourceQueue.length > 0) {
```

Comment:
Maybe we can use a for in loop and not use 2 different list to check whether an element has been itereated over:

```js
for (let i; i<resourceQueue.length-1;i++){
// don't shift the array

resourceQueue = (...new Set([...resourceQueue, ...nestedReferences]))
} 

```

Set preserve order => https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set#description

Reply:
Agree that it isn't clean to check 2 different lists.

But still a bit hesitant about this alternative. A queue makes more sense than a for loop since `resourceQueue` will have a varying length throughout the code execution (might make it more confusing).

As discussed, will make use of a separate Set object to keep track of everything that has been fetched.